File: src/main/java/net/pterodactylus/sone/web/EditImagePage.java
/*
 * Sone - EditImagePage.java - Copyright © 2010–2015 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web;

import net.pterodactylus.sone.data.Image;
import net.pterodactylus.sone.text.TextFilter;
import net.pterodactylus.sone.web.page.FreenetRequest;
import net.pterodactylus.util.template.Template;
import net.pterodactylus.util.template.TemplateContext;
import net.pterodactylus.util.web.Method;

/**
 * Page that lets the user edit title and description of an {@link Image}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class EditImagePage extends SoneTemplatePage {

	/**
	 * Creates a new “edit image” page.
	 *
	 * @param template
	 *            The template to render
	 * @param webInterface
	 *            The Sone web interface
	 */
	public EditImagePage(Template template, WebInterface webInterface) {
		super("editImage.html", template, "Page.EditImage.Title", webInterface, true);
	}

	//
	// SONETEMPLATEPAGE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected void processTemplate(FreenetRequest request, TemplateContext templateContext) throws RedirectException {
		super.processTemplate(request, templateContext);
		if (request.getMethod() == Method.POST) {
			String imageId = request.getHttpRequest().getPartAsStringFailsafe("image", 36);
			String returnPage = request.getHttpRequest().getPartAsStringFailsafe("returnPage", 256);
			Image image = webInterface.getCore().getImage(imageId, false);
			if (image == null) {
				throw new RedirectException("invalid.html");
			}
			if (!image.getSone().isLocal()) {
				throw new RedirectException("noPermission.html");
			}
			if ("true".equals(request.getHttpRequest().getPartAsStringFailsafe("moveLeft", 4))) {
				image.getAlbum().moveImageUp(image);
			} else if ("true".equals(request.getHttpRequest().getPartAsStringFailsafe("moveRight", 4))) {
				image.getAlbum().moveImageDown(image);
			} else {
				String title = request.getHttpRequest().getPartAsStringFailsafe("title", 100).trim();
				String description = request.getHttpRequest().getPartAsStringFailsafe("description", 1024).trim();
				if (title.length() == 0) {
					throw new RedirectException("emptyImageTitle.html");
				}
				image.modify().setTitle(title).setDescription(TextFilter.filter(request.getHttpRequest().getHeader("host"), description)).update();
			}
			webInterface.getCore().touchConfiguration();
			throw new RedirectException(returnPage);
		}
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/pages/EditImagePage.kt
package net.pterodactylus.sone.web.pages

import net.pterodactylus.sone.data.*
import net.pterodactylus.sone.data.Image.Modifier.*
import net.pterodactylus.sone.main.*
import net.pterodactylus.sone.text.*
import net.pterodactylus.sone.utils.*
import net.pterodactylus.sone.web.*
import net.pterodactylus.sone.web.page.*
import net.pterodactylus.util.template.*
import javax.inject.*

/**
 * Page that lets the user edit title and description of an {@link Image}.
 */
class EditImagePage @Inject constructor(webInterface: WebInterface, loaders: Loaders, templateRenderer: TemplateRenderer) :
		LoggedInPage("editImage.html", "Page.EditImage.Title", webInterface, loaders, templateRenderer) {

	override fun handleRequest(soneRequest: SoneRequest, currentSone: Sone, templateContext: TemplateContext) {
		if (soneRequest.isPOST) {
			val image = soneRequest.core.getImage(soneRequest.httpRequest.getPartAsStringFailsafe("image", 36)) ?: throw RedirectException("invalid.html")
			if (!image.sone.isLocal) {
				throw RedirectException("noPermission.html")
			}
			soneRequest.httpRequest.getPartAsStringFailsafe("returnPage", 256).let { returnPage ->
				if (soneRequest.httpRequest.getPartAsStringFailsafe("moveLeft", 4) == "true") {
					image.album.moveImageUp(image)
					soneRequest.core.touchConfiguration()
				} else if (soneRequest.httpRequest.getPartAsStringFailsafe("moveRight", 4) == "true") {
					image.album.moveImageDown(image)
					soneRequest.core.touchConfiguration()
				} else {
					try {
						image.modify()
								.setTitle(soneRequest.httpRequest.getPartAsStringFailsafe("title", 100))
								.setDescription(TextFilter.filter(soneRequest.httpRequest.getHeader("Host"), soneRequest.httpRequest.getPartAsStringFailsafe("description", 1024)))
								.update()
						soneRequest.core.touchConfiguration()
					} catch (e: ImageTitleMustNotBeEmpty) {
						throw RedirectException("emptyImageTitle.html")
					}
				}
				throw RedirectException(returnPage)
			}
		}
	}

}
